Sources/NIOHTTP2/HTTP2StreamChannel.swift

```diff
@@ -599,4 +643,59 @@ internal extension HTTP2StreamChannel {
     func receiveParentChannelReadComplete() {
         self.tryToRead()
     }
+
+    func parentChannelWritabilityChanged(newValue: Bool) {
+        // There's a trick here that's worth noting: if the child channel hasn't either sent a frame
+        // or been activated on the network, we don't actually want to change the observable writability.
+        // This is because in this case we really want user code to send a frame as soon as possible to avoid
+        // issues with their stream ID becoming out of date. Once the state transitions we can update
+        // the writability if needed.
+        guard case .changed(newValue: let localValue) = self.writabilityManager.parentWritabilityChanged(newValue) else {
+            return
+        }
+
+        // Ok, the writability changed.
+        switch self.state {
+        case .idle, .localActive:
+            // Do nothing here.
+            return
+        case .remoteActive, .active, .closing, .closingNeverActivated, .closed:
+            self._isWritable.store(localValue)
+            self.pipeline.fireChannelWritabilityChanged()
+        }
+    }
+}
+
+extension HTTP2Frame {
+    /// A shorthand heuristic for how many bytes we assume a frame consumes on the wire.
+    ///
+    /// Here we concern ourselves only with per-stream frames: that is, `HEADERS`, `DATA`,
+    /// `WINDOW_UDPATE`, `RST_STREAM`, and I guess `PRIORITY`. As a simple heuristic we
+    /// hard code fixed lengths for fixed length frames, use a calculated length for
+    /// variable length frames, and just ignore encoded headers because it's not worth doing a better
+    /// job.
+    fileprivate var bufferBytes: Int {
+        switch self.payload {
+        case .data(let d):
+            let paddingBytes = d.paddingBytes.map { $0 + 1 } ?? 0
+            return d.data.readableBytes + paddingBytes + 9
+        case .headers(let h):
+            let paddingBytes = h.paddingBytes.map { $0 + 1 } ?? 0
+            return paddingBytes + 9
+        case .priority:
+            return 9 + 5
+        case .pushPromise(let p):
+            // Like headers, this is variably size, and we just ignore the encoded headers because
+            // it's not worth having a heuristic.
+            let paddingBytes = p.paddingBytes.map { $0 + 1 } ?? 0
+            return paddingBytes + 9
+        case .rstStream:
+            return 9 + 4
+        case .windowUpdate:
+            return 9 + 4
+        default:
+            // Unknown or unexpected control frame: say 9 bytes.
+            return 9
```

Comment:
as we use 9 here multiple times I think it would be good to make this a static let or something like this and give it a meaningful name (and so also give a hind where this is coming from, like the "frame header" size).